aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -103,18 +108,45 @@ compute_result_type(at::ArrayRef<OperandInfo> operands,
 
 std::tuple<Device, ScalarType> TensorIterator::compute_common_type() {
   // See [Result type computation] in TensorIterator.h
+
   auto result_type =
       compute_result_type(operands_,
-        [](const Tensor& t) { return t.dim() > 0; },
-        [](const Tensor& t) { return !t.unsafeGetTensorImpl()->is_wrapped_number(); },
-        [](const Tensor& t) { return true; });
+        [](const OperandInfo& op) { return op.tensor.dim() > 0; },
+        [](const OperandInfo& op) { return !op.tensor.unsafeGetTensorImpl()->is_wrapped_number(); },
+        [](const OperandInfo& op) { return true; });
+
+  // if non-zero-dim tensor result is an integral type and there's a zero-dim
+  // floating point operand, we'll promote the the floating point type.
+  if (isIntegralType(std::get<1>(result_type))) {
+    auto alternate = compute_result_type(operands_,
+        [](const OperandInfo& op) {
+          return isFloatingType(op.tensor.scalar_type()) && op.tensor.dim() == 0;
+        }
+    );
+    if (std::get<1>(alternate) != ScalarType::Undefined) {
+      // preserve device from original result
+      return std::make_tuple(std::get<0>(result_type), std::get<1>(alternate));
+    }
+  }
 
   TORCH_INTERNAL_ASSERT(std::get<1>(result_type) != ScalarType::Undefined);
   return result_type;
 }
 
+static bool can_cast(const ScalarType& from, const ScalarType& to) {
+  // we disallow float -> integral, e.g., float_tensor *= int is disallowed.
```

Comment:
one thing I asked in the previous PR is that we actually document the intent of what kind of casting we are doing here.

For example, I don't believe the following matches numpy's same_kind:
```
>>> np.can_cast(np.uint8, np.bool, casting='same_kind')
False
```

Reply:
Do you have a preference on where that would go? maybe tensor_attributes.rst? Or would code comments be sufficient?

also this code comment was wrong, was supposed to say int_tensor *= float disallowed.

Reply:
updated the code comment to be correct and added an explanation in tensor_attributes.rst.